src/keyboardNavigation.ts

```diff
@@ -0,0 +1,83 @@
+import { isDeepestVisible, isImmediateParent } from './utils';
+
+function findDeepestVisibleChild(node: HTMLElement): HTMLElement {
+    if (isDeepestVisible(node)) return node;
+    const ul = <HTMLElement>node.children[1];
+    return findDeepestVisibleChild(<HTMLElement>ul.children[ul.children.length - 1]);
+}
+
+function findNearestParentWithAdditionalLeaves(node: HTMLElement): HTMLElement {
+    const ul = node.parentElement;
+    if (!ul) return node;
+    const parent = ul.parentElement;
+    if (!parent || !parent.parentElement) return node;
+    const parentAndSiblings = [...parent.parentElement.children];
+    if (parentAndSiblings.indexOf(parent) == parentAndSiblings.length - 1) {
+        return findNearestParentWithAdditionalLeaves(parent);
+    } else {
+        return <HTMLElement>parentAndSiblings[parentAndSiblings.indexOf(parent) + 1];
+    }
+}
+
+export const navigateUp = (path: string, shadowRoot: ShadowRoot): void => {
```

Comment:
I feel like most of the navigate functions could benefit from extra helper functions to make parsing the function easier. What do you think about maybe adding some helpers like `getChildNode`, `getNextSibling`, `isLastChild`

Something else (minor) to consider would be tweaking the `if/else` to just assign a new var instead of flow control, e.g

```js
    const newFocus = isDeepestVisible(currentFocus)
        ? getChildNode(childNode) 
            : isLastChild(currentFocusAndSiblings, currentFocus)
        ? findNearestParentWithAdditionalLeaves(currentFocus) 
            : getNextSibling(currentFocusAndSiblings, currentFocus);

    if (newFocus) changeFocusedElement(newFocus, currentFocus)
```

I tend to find it easier to reason about, the flow control only applies to the element to focus rather than what function to call.

Reply:
Definitely agree that these benefit from some helpers. I added a collection that made reading these functions easier. 
Changes applied 



I took a swing at the ternary stuff. I find it harder to read, particularly with how complex `navigateDown` is. The formatting provided by the repo's formatter is also not helping me. Do you think its better this way? I'm not opposed to going this direction if you think it is, but I would revert if up to me. 
Changes applied (for now at least)

Reply:
Granted, a lot of these stylistic preferences may be moot depending on the repo owner's preferences lol

Reply:
True! I went ahead and opted out of the `navigateDown` ternary and went for a locally scoped function. I was trying to adjust some logic in that ternary and was not having a good time